### PR TITLE
posix: Clean up various headers

### PR DESCRIPTION
**Pull request by pfalcon**
Ok, so this started as just moving open() from unistd.h to fcntl.h, but as usual, led to chain of conflicts, which however this time were tracked down to completion.

---
Per POSIX, open() is defined in <fcntl.h>. fcntl.h in turn comes from
the underlying libc, either newlib, or minimal libc.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

**Comment by pfalcon** [comment]
Ok, so this started as just moving open() from unistd.h to fcntl.h, but as usual, led to chain of conflicts, which however this time were tracked down to completion.

File: lib/libc/minimal/include/sys/stat.h
Comment (mbolivar) [inline lib/libc/minimal/include/sys/stat.h]: Why keep this? It's not required for the definitions later in the header, and seems a bit of a layering violation for the C library to be pulling in Zephyr kernel headers.

File: include/posix/time.h
Comment (mbolivar) [inline include/posix/time.h]: Have you tested with the recently supported newlib-nano as well?

File: include/posix/time.h
Comment (mbolivar) [inline include/posix/time.h]: If this is just for XTensa, why not ifndef under that arch too, so other arches for sure aren't negatively impacted by unexpected changes here?

**Review by mbolivar — DISMISSED** [review]
If the goal is a cleanup, I think at least including kernel.h from a libc header ought to be addressed, as it does fall under that umbrella.

**Comment by pfalcon** [comment]
@mbolivar: I don't understand what you mean. Please review previous comments and history of title changes regarding the goal of this PR. And perhaps, post comments without hitting -1 button right away?

**Comment by mbolivar** [comment]
>  post comments without hitting -1 button right away?

I posted 3 comments; what are you talking about?

**Comment by pfalcon** [comment]
> I posted 3 comments; what are you talking about?

Ok, sorry, I didn't get them in email, and didn't see at once here.

File: lib/libc/minimal/include/sys/stat.h
Comment (pfalcon) [reply lib/libc/minimal/include/sys/stat.h]: > Why keep this?

In general, to not introduce potential regressions. If the change says "move file", it should move file with minimal changes (I had to update header guards, as you imagine), instead of trying to squeeze other "helpful" stuff, which someone later may debug, pulling their hair out.

> and seems a bit of a layering violation for the C library to be pulling in Zephyr kernel headers.

Well, for an I/O related header like sys/stat.h there's nothing unusual to include headers related to OS I/O implementation (like kernel headers). Who knows, maybe something which includes this header relies later on those definitions to be available. Issue that or not, that's not what I fix here. Again, this started as a simple 1-line patch of moving 1 line from one file to another. This caused 4 levels of cascading failures, and turned out to required 5 patch to make CI pass. There's nothing more here to fix in *this PR*. (And a lot to fix in many more PRs.)

File: include/posix/time.h
Comment (pfalcon) [reply include/posix/time.h]: > 

Have you tested with the recently supported newlib-nano as well?


Recently supported? Where? By whom?

File: include/posix/time.h
Comment (pfalcon) [reply include/posix/time.h]: > If this is just for XTensa,

And what if this not only by Xtensa, and later we have another FooBarArch with the same issue? This kludge tests for the actual indication of the problem - very old Newlib version, and is consistent to previous such workaround already in the codebase.

Comment (mbolivar) [reply include/posix/time.h]: Then it's not just for XTensa. Perhaps the comment could be clarified by changing "for XTensa" to "for very old Newlib versions" or so?

File: include/posix/time.h
Comment (mbolivar) [reply include/posix/time.h]: https://github.com/zephyrproject-rtos/zephyr/commit/472ea92e79e346ad390ebbfc1ff27a5776f4fa25

File: lib/libc/minimal/include/sys/stat.h
Comment (mbolivar) [reply lib/libc/minimal/include/sys/stat.h]: > In general, to not introduce potential regressions. 

Sounds to me like any such code not caught by sanitycheck was buggy in the first place by using things it couldn't know for sure were included.

Why not remove it in a separate commit in the same PR and see what sanitycheck says?

Comment (mbolivar) [reply lib/libc/minimal/include/sys/stat.h]: > not caught by sanitycheck

And out of tree users will have to do their own testing, of course. If I previously had code that pulled in fcntl.h and relied that to pull in zephyr's gpio.h, and someone changed things around so my build breaks, I think that's my fault, not theirs.

File: include/posix/time.h
Comment (pfalcon) [reply include/posix/time.h]: I hope I explained why I did like that: a) trying to, and usually go for, more general solution; b) yet explaining the motivation why ugliness like that appeared there at all.

You really think there's anything to change there (besides blacklisting Xtensa from POSIX CI tests)? Because I humbly think there's nothing to change for good, only for the purpose of bikeshedding.

File: include/posix/time.h
Comment (pfalcon) [reply include/posix/time.h]: > 472ea92

Ok, I'm supporting only what Zephyr CI supports. (Because I have to, or my patches won't have a chance to be merged, even though the right solution in various cases is to blacklist undermaintained platforms from CI and wait till their maintainers catch up with the rest of Zephyr).

File: lib/libc/minimal/include/sys/stat.h
Comment (pfalcon) [reply lib/libc/minimal/include/sys/stat.h]: > Sounds to me like any such code not caught by sanitycheck was buggy in the first place by using things it couldn't know for sure were included.

... And yet I still can't fix all the wrongs on Zephyr in one PR, instead trying to do that PR by PR. The sooner one PR merged, the sooner more fixes come in. Anyway, back to topic,  which color do we like them, bikesheds? Me likes green.

> Why not remove it in a separate commit in the same PR and see what sanitycheck says?

I can do that by personal request of yours, and with full understanding that it's outside the scope of this PR ;-).

File: include/posix/time.h
Comment (mbolivar) [reply include/posix/time.h]: I think that's a fine answer.

File: include/posix/time.h
Comment (mbolivar) [reply include/posix/time.h]: The comment says "for XTensa". While reviewing it, I wondered why it wasn't siloed away with ifdefs for that architecture only, especially to avoid unexpected changes on other architectures that are playing by the rules.

When I asked about that, you explained it's not just for XTensa. So I asked why not make the comment more accurate, which would have taken you about as much time to do as it did to write your reply.

You accused me of bikeshedding.

Have a nice day.

**Comment by mbolivar** [comment]
I'm done with this PR. Dismissing my stale review.

File: lib/libc/minimal/include/sys/stat.h
Comment (pabigot) [inline lib/libc/minimal/include/sys/stat.h]: I'm with @mbolivar from the perspective of minimal libc headers should not be including kernel.h.  However I'll say it shouldn't have been put into the Zephyr's POSIX material in the first place, so that ship has sailed and though removing the include should happen it's not in scope for this PR.

**Review by pabigot — APPROVED** [review]

**Comment by pabigot** [comment]
I would like to see this merged soon since it's required as the basis of reworking #17003.

**Comment by pfalcon** [comment]
@pabigot: Thanks for approval.

File: lib/libc/minimal/include/sys/stat.h
Comment (nashif) [reply lib/libc/minimal/include/sys/stat.h]: Should not be including kernel.h here

**Review by nashif — DISMISSED** [review]

File: lib/libc/minimal/include/sys/stat.h
Comment (pfalcon) [reply lib/libc/minimal/include/sys/stat.h]: > Should not be including kernel.h here

Again, I just moved the file around, and made minimal change to accompany that, if any. Let me try to remove that and see if it leads to cascade of unexpected CI breaks (which is very common so far with POSIX subsys stuff, and what I'm trying to fix - step by step).

**Comment by pfalcon** [comment]
Rebased to the latest master and removed `#include <kernel>` from sys/stat.h to accommodate thorough reviewers here.

**Comment by pfalcon** [comment]
> Let me try to remove that and see if it leads to cascade of unexpected CI breaks (which is very common so far with POSIX subsys stuff, and what I'm trying to fix - step by step).

Failed. Surprise.

But this time it looks like some regression from timer (counter?) API rework. I'm curious why main CI didn't/doesn't catch it  - I also see our (Linaro's) CI fails on that. (Didn't have a chance to look closely, as we're at a sprint now.)

**Comment by pfalcon** [comment]
~~~
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/apic_timer.c:6:10: fatal error: drivers/system_timer.h: No such file or directory
 #include <drivers/system_timer.h>
~~~

**Comment by pfalcon** [comment]
Build is green, @nashif, please re-review.

**Comment by pabigot** [comment]
@pfalcon this needs a rebase so we can push it toward a merge.  Thanks.

**Comment by pfalcon** [comment]
> this needs a rebase

Rebased, thanks. The build is green. (Note that I'm going on vacation for a week re: any possible future requests.)

**Comment by pabigot** [comment]
@pfalcon Thanks.   My rework of #17003 is complete and waiting for this to be merged so I can submit those changes and DS3231 support.  I'm hoping that'll happen soon.  Enjoy your vacation.

**Comment by andrewboie** [comment]
@nashif can you re-visit your review?

**Review by tyrantlizard77 — APPROVED** [review]

**Review by mgielda — APPROVED** [review]
LGTM

**Comment by mgielda** [comment]
(I know Anas is on holiday and double checked that his remark was included, so I dismissed it - hope that's OK)

**Comment by pfalcon** [comment]
@mgielda, Thanks for review.


Also, the PR is rebased against master, added a commit to fix issue mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/17768#issuecomment-515015980 (cc: @PiotrZierhoffer).

File: lib/libc/minimal/include/time.h
Comment (galak) [inline lib/libc/minimal/include/time.h]: How does this work, where is <sys/_timespec.h> coming from?

File: include/posix/time.h
Comment (galak) [inline include/posix/time.h]: Can you add a comment of which case is handling the old xtensa version

File: lib/libc/minimal/include/time.h
Comment (galak) [reply lib/libc/minimal/include/time.h]: (maybe the commit order is being shown oddly by github)

File: lib/libc/minimal/include/sys/stat.h
Comment (galak) [inline lib/libc/minimal/include/sys/stat.h]: I'd keep the Intel Copyright and just add the Linaro one.

**Review by galak — CHANGES_REQUESTED** [review]
A few minor comments, but otherwise looks good.

File: lib/libc/minimal/include/sys/stat.h
Comment (pfalcon) [reply lib/libc/minimal/include/sys/stat.h]: > I'd keep the Intel Copyright and just add the Linaro one.

Ok. Note that the original sys/stat.h was an empty file with just a comment "Dummy sys/stat.h to fulfill the requirements of certain libraries."

File: lib/libc/minimal/include/time.h
Comment (pabigot) [reply lib/libc/minimal/include/time.h]: @galak https://github.com/isaacs/github/issues/386#issuecomment-494288396

File: lib/libc/minimal/include/sys/stat.h
Comment (galak) [reply lib/libc/minimal/include/sys/stat.h]: Sure, just felt like we are moving most of the contents of include/posix/sys/stat.h into there.

Comment (pfalcon) [reply lib/libc/minimal/include/sys/stat.h]: > I'd keep the Intel Copyright and just add the Linaro one.

D'oh, so what happened here, is that we had a dummy, empty file copyrighted Linaro. It gets overwritten by a moved file copyrighted by Intel. I didn't specifically change any copyright, just one file gets completely overwritten by another.

@galak, I hope you agree that under such circumstances, it doesn't make sense to stamp Linaro's copyright there.

File: include/posix/time.h
Comment (pfalcon) [reply include/posix/time.h]: > Can you add a comment of which case is handling the old xtensa version

Done.

File: lib/libc/minimal/include/sys/stat.h
Comment (pfalcon) [reply lib/libc/minimal/include/sys/stat.h]: Ok, so hopefully this can be closed.

**Comment by pfalcon** [comment]
> A few minor comments, but otherwise looks good.

Should be addressed now.

File: lib/libc/minimal/include/sys/stat.h
Comment (galak) [reply lib/libc/minimal/include/sys/stat.h]: I'm fine not having the linaro copyright.

**Review by galak — APPROVED** [review]

**Pull request merged**
